lightning/src/ln/onion_utils.rs

```diff
@@ -2107,6 +2197,160 @@ const HMAC_COUNT: usize = MAX_HOPS * (MAX_HOPS + 1) / 2;
 
 pub const ATTRIBUTION_DATA_LEN: usize = MAX_HOPS * HOLD_TIME_LEN + HMAC_COUNT * HMAC_LEN;
 
+/// Adds the current node's hmacs for all possible positions to this packet.
+pub(crate) fn add_hmacs(
+	shared_secret: &[u8], message: &[u8], packet: &mut [u8; ATTRIBUTION_DATA_LEN],
+) {
+	let (hold_times, hmacs) = packet.split_at_mut(MAX_HOPS * HOLD_TIME_LEN);
+
+	let um: [u8; 32] = gen_um_from_shared_secret(&shared_secret);
+
+	// Iterate over all possible positions that this hop could be on the path. An intermediate node does not have this
+	// information, so it is up to the sender to verify the HMAC that corresponds to the actual position.
+	for hmac_idx in 0..MAX_HOPS {
+		// Calculate position relative to the final node. The final node is at position 0.
+		let position: usize = MAX_HOPS - hmac_idx - 1;
+
+		// The hmac covers the original message and - for the assumed position - all the hold times and downstream
+		// hmacs.
+		let mut hmac_engine = HmacEngine::<Sha256>::new(&um);
+		hmac_engine.input(&message);
+		hmac_engine.input(&hold_times[..(position + 1) * HOLD_TIME_LEN]);
+		write_downstream_hmacs(position, hmacs, &mut hmac_engine);
+
+		let full_hmac = Hmac::from_engine(hmac_engine).to_byte_array();
+
+		// Truncate the hmac to save space. A low-probability collision acceptable here because the consequence is just
+		// a pathfinding penalty.
+		let hmac = &full_hmac[..HMAC_LEN];
+
+		// Store the new hmac.
+		hmacs[hmac_idx * HMAC_LEN..(hmac_idx + 1) * HMAC_LEN].copy_from_slice(hmac);
+	}
+}
+
+/// Writes the HMACs corresponding to the given position that have been added already by downstream hops. Position is
+/// relative to the final node. The final node is at position 0.
+pub fn write_downstream_hmacs(position: usize, hmacs: &[u8], w: &mut HmacEngine<Sha256>) {
```

Comment:
Reading through these, I do wonder whether defining a type for the attribution data would be helpful, something like:

```

type HoldTime = [u8; HOLD_TIME_LEN];
type TruncatedHMAC = [u8; HMAC_LEN];

struct AttributionData {
	hold_times: [HoldTime; HOLD_TIME_LEN],
	hmacs: [TruncatedHMAC; HMAC_COUNT],
}
```

Benefit being that it allows us to think about these element in their appropriate "chunks" of individual hold times or hmacs, rather than also needing to worry about length of the underlying bytes.

Reply:
I've thought about this too, but was discouraged a bit because writing out the whole array in the wire message etc might become more complicated. Will take another look, because there's a lot of slicing required now.

Reply:
I gave this a try in https://github.com/lightningdevkit/rust-lightning/commit/11d548592e0cf6d983d9fe6c3f98130aa4562c7c

Straight-forward, but the open question is how define the wire message. Is it correct to also there use this new struct? Can't use the nested tlv serialization, so would need to join the two fields there somehow. If the wire message would have the single concatenated field, then the question becomes where to do the concatenation of the separate fields from `OnionErrorPacket`. That gets spread out a bit over the code base I think.

Reply:
Ok, I think I can just do it with non-tlv serialization at it might just all work.

Reply:
> https://github.com/lightningdevkit/rust-lightning/compare/main...joostjager:rust-lightning:attr-errs-split-attr-data

This looks reasonable to me!

> the open question is how define the wire message

Also not sure about this, somebody with more codebase familiarity would need to comment. 

My understanding is that `msgs` is supposed to represent the structure of the message on the wire as closely as possible, which `AttributableData` doesn't do, but it does seem worth it to have the it there so we don't have to convert all over the place.

Reply:
I've now pushed a version with the arrays split, but still keeping them u8 arrays. It's not the biggest change, mainly saving some .split / .splt_mut calls, but I think an improvement nonetheless.

I also tried using stronger types (u32 for hold time and [u8; 4] for truncated hmac), but I ran into the chacha processing which takes a u8 array. Maybe there's an (unsafe) rust way to do it? Not sure if it is worth it.